sdk/storage/azblob/blockblob/chunkwriting.go

```diff
@@ -12,225 +12,306 @@ import (
+		desiredVal, morphResult := morpher(currentVal)
+		if atomic.CompareAndSwapUint32(target, currentVal, desiredVal) {
+			return morphResult
+		}
 	}
+}
 
-	if cerr := c.getErr(); cerr != nil {
-		return cerr
-	}
+type blockID [64]byte
 
-	return err
+func (blockID blockID) ToBase64() string {
+	return base64.StdEncoding.EncodeToString(blockID[:])
 }
 
-// write uploads a chunk to blob storage.
-func (c *copier) write(chunk copierChunk) {
-	defer c.o.transferManager.Put(chunk.buffer)
+type uuidBlockID blockID
 
-	if err := c.ctx.Err(); err != nil {
-		return
-	}
-	stageBlockOptions := c.o.getStageBlockOptions()
-	_, err := c.to.StageBlock(c.ctx, chunk.id, shared.NopCloser(bytes.NewReader(chunk.buffer[:chunk.length])), stageBlockOptions)
-	if err != nil {
-		select {
-		case c.errCh <- err:
-			// failed to stage block, cancel the copy
-		default:
-			// don't block the goroutine if there's a pending error
-		}
-	}
+func newUUIDBlockID(u uuid.UUID) uuidBlockID {
+	ubi := uuidBlockID{}     // Create a new uuidBlockID
+	copy(ubi[:len(u)], u[:]) // Copy the specified UUID into it
+	// Block number defaults to 0
+	return ubi
 }
 
-// close commits our blocks to blob storage and closes our writer.
-func (c *copier) close() error {
-	c.wg.Wait()
-
-	if err := c.getErr(); err != nil {
-		return err
-	}
+func (ubi uuidBlockID) WithBlockNumber(blockNumber uint32) uuidBlockID {
+	binary.BigEndian.PutUint32(ubi[len(uuid.UUID{}):], blockNumber) // Put block number after UUID
+	return ubi                                                      // Return the passed-in copy
+}
 
-	var err error
-	commitBlockListOptions := c.o.getCommitBlockListOptions()
-	c.result, err = c.to.CommitBlockList(c.ctx, c.id.issued(), commitBlockListOptions)
-	return err
+func (ubi uuidBlockID) ToBase64() string {
+	return blockID(ubi).ToBase64()
 }
 
-// id allows the creation of unique IDs based on UUID4 + an int32. This auto-increments.
-type id struct {
-	u   [64]byte
-	num uint32
-	all []string
+// mmbPool implements the bufferManager interface.
+// it uses anonymous memory mapped files for buffers.
+// don't use this type directly, use newMMBPool() instead.
+type mmbPool struct {
+	buffers chan mmb
+	count   int
+	max     int
+	size    int64
 }
 
-// newID constructs a new id.
-func newID(uu uuid.UUID) *id {
-	u := [64]byte{}
-	copy(u[:], uu[:])
-	return &id{u: u}
+func newMMBPool(maxBuffers int, bufferSize int64) bufferManager[mmb] {
+	return &mmbPool{
+		buffers: make(chan mmb, maxBuffers),
+		max:     maxBuffers,
+		size:    bufferSize,
+	}
 }
 
-// next returns the next ID.
-func (id *id) next() string {
-	defer atomic.AddUint32(&id.num, 1)
+func (pool *mmbPool) Acquire() <-chan mmb {
+	return pool.buffers
+}
 
-	binary.BigEndian.PutUint32(id.u[len(uuid.UUID{}):], atomic.LoadUint32(&id.num))
-	str := base64.StdEncoding.EncodeToString(id.u[:])
-	id.all = append(id.all, str)
+func (pool *mmbPool) Grow() (int, error) {
+	if pool.count < pool.max {
```

Comment:
This doesn't need to be thread safe because Grow is only ever called from the 1 goroutine calling UploadStream, right?

Reply:
Correct.  I've added a comment to explain this.